### ISSUE TITLE
Tutorial on how to train a custom YoloV8 model without skipping steps

### ISSUE THREAD

**Issue by mgabell** [ultralytics/ultralytics#5335]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I would like if someone could confirm if the following steps to train a custom model with ultralytics is correct. I hope also that
someone could comment on my questions I have in the following "tutorial"

I understand the following but neither of the creators create a good match of packages so that this can be avoided:
To run training protobuf must be 3.19.1.
To convert to onnx, the latest pytorch, torchvision and onnx must be installed and protobuf upgraded to latest. When training again the protobuf must be downgraded to 3.19.1 again.

"Custom model creation":
1. Install requirements according to ultralytics.
2. Create a folder called datasets in ./ultralytics where the training of images with the accompanying txt file for each image. With class number, and box location. 
`3 0.72283 0.37127 0.08152000000004 0.23508`
This should have the same filename as the image except .txt as file extension.
4. Create a validation set of images with a similar setup.
5. Create a folder inside ./ultralytics called yaml (for convenience)
6. Create a file called mytrainingset.yaml and write the following and update according to your need:
```
path: ./datasets
train: ./train
val: ./val
test:

#number of classes (Those you have in your training set)
2

#Class names
names: 0: ball, 1: stones

```
7. Save the file
8. in the ultralytics folder create a train.py file and save it with the following content.

```
from ultralytics import YOLO

# Load a model
model = YOLO("yolov8s.pt")  # load a pretrained model. change this to the model you want to use, n, s, l etc. (only ending)

# Train the model
results = model.train(data="./yaml/mytrainingset.yaml", imgsz=800, batch=8, epochs=50, plots=False)

```

> Comment: So yo my understanding. The code will load a pre-trained yolov8 model of type small. Then all needed is to train it and it will create a best.pt under results with the new updated model. There is no need at this stage to add. create or download any cfg files or weight files?

9. To run inference using the new model we create a new python file:

```
from ultralytics import YOLO
#This is the location of the newly trained model, based on yolov8s.pt. The folder change name when training again, called train2, #train3 etc...
model = YOLO("./runs/detect/train/weights/best.pt") 
#The save-text save a txt file with the inference results, box and class, whilst save also saves image. In this case it does not save 
#image...
results = model.predict(source="./my_images_that_I_want_to_run_inference_on/", save-text=True, save=False)
```

**Now to the tricky part of running deepstream on AGX Orin**
10. Git clone https://github.com/marcoslucianops/DeepStream-Yolo/tree/master to ~/ or the one above ./ultralytics.

> This is where the information gets scattered. This is not straight forward since the different versions of Deepstream-Yolo and ultralytics and deepstream-app from nVidia does not add up in terms of library versions etc.

1. Install Jetpack of prefered choice. I think I have JP5.5
Convert the yolov8 pt file to ONNX. This is shady. According to seeed:
https://wiki.seeedstudio.com/YOLOv8-DeepStream-TRT-Jetson

2. Download Deepstream-Yolo to home catalogue: git clone https://github.com/marcoslucianops/DeepStream-Yolo
Two ways of doing this. If you go with the latest version you need to follow above instruction to the letter. But seeed does the following:
 
3. cd DeepStream-Yolo
4. git checkout 68f762d5bdeae7ac3458529bfe6fed72714336ca

6. cp utils/gen_wts_yoloV8.py ~/ultralytics

According to https://wiki.seeedstudio.com/YOLOv8-DeepStream-TRT-Jetson you are supposed to create the weights and cfg file. But it is not needed from the latest commit. So depending on the version there are two ways...

> NOTE: You can use your custom model, but it is important to keep the YOLO model reference (yolov8_) in your cfg and weights/wts filenames to generate the engine correctly. 

What does it mean??? Which should we use. How does this relate to the onnx file? Can I use the onnx instead of weights and cfg?

`python3 gen_wts_yoloV8.py -w yolov8s.pt`
This is already not working because ultralytics.yolo does not exist and is instead ultralytics.XXX so remove .yolo from the get_wts_yoloV8.py

7. Copy the files to DeepStream-Yolo:
cp yolov8s.cfg ~/DeepStream-Yolo
cp yolov8s.wts ~/DeepStream-Yolo
cp labels.txt ~/DeepStream-Yolo

8. Now its time to build the nvdsinfer_custom_impl_Yolo  "with the weights (I assume)"
cd ~/DeepStream-Yolo
CUDA_VER=11.4 make -C nvdsinfer_custom_impl_Yolo  # for DeepStream 6.2/ 6.1.1 / 6.1

So after this I would have the nvdsinfer_custom_impl_Yolo  built with the weights and cfg from the newly trained custom yolov8 model. This would then be used in the nvdsinfer_custom_impl_Yolo when building. So that when I run inference with my ONNX file I have the correct weights and cfg. That would mean that I have to recompile the nvdsinfer_custom_impl_Yolo every time I update my model.

> So the relation between the nvdsinfer_custom_impl_Yolo and my cfg/weights files is the building of the file. But how is the ONNX using it. You are not supposed to update the config file with weights and cfg if using onnx. You only reference the ONNX and nvdsinfer_custom_impl_Yolo. Is that the difference? If using onnx its in nvdsinfer_custom_impl_Yolo, and if not you reference the weights and cfg?

```
For weights and cfg:
```
[property]
...
custom-network-config=yolov8s.cfg
model-file=yolov8s.wts
...
num-detected-classes=80
```
for ONNX:
[property]
...
onnx-file=yolov8s.onnx
...
num-detected-classes=80
...
parse-bbox-func-name=NvDsInferParseYolo

```
..........

If using onnx file it should be converted by:
1. Copy export_yolov8.py to ./ultralytics.
2. python3 export_yoloV8.py -w yolov8s.pt --dynamic

> COMMENT: Do I need the weights and cfg file here??

> **Why do I write all this?**
> I can not get my newly created best.pt based on yolov8s.pt with 15000 images to work properly when converting to onnx with 
> python3 export_yoloV8.py -w yolov8s.pt --dynamic. Using the YoloV8 custom model with deepstream and the deepstream-test1.py from nVidia does not create any inference boxes. Sometimes it seems so, but that is when I think the cfg and weights files are from the original yolov8.pt. So I assume I have not the propert weights and cfg somewhere? 
> Also to mention that training with ultralytics require protobuf 3.19.1 and converting to onnx uses latest protobuf. why? Because there is a difference between export_yolov8.py and training the model with pytorch/torchvision and onnx. There is not a proper documentation that works across the different solutions from:
> https://github.com/marcoslucianops/DeepStream-Yolo/blob/master/docs/YOLOv8.md
> and 
> https://github.com/ultralytics/ultralytics/tree/main

**Comment by mgabell** [comment]
![image](https://github.com/ultralytics/ultralytics/assets/78966452/bb91b4a6-e4f4-4c60-ba2c-c3a5b85586e3)

**Comment by zshariff6506** [comment]
Howdy!

Honestly you first part of the project looks fine and impressive! If you are doing 15000 images and detect, think of moving to yolov8m rather small.

The second part with AGN Onix, and I don't understand from that point.

> Comment: So yo my understanding. The code will load a pre-trained yolov8 model of type small. Then all needed is to train it and it will create a best.pt under results with the new updated model. There is no need at this stage to add. create or download any cfg files or weight files?

**Yeah, after training with your data using yolov8s.pt, you achieve a best.pt which carries cfg and weights. As mentioned, consider yolov8m**!

Hope this helps!

**Comment by glenn-jocher** [comment]
@mgabell hello!

You've grasped it correctly! When you train using the pretrained YOLOv8 model (like yolov8s.pt), it will generate a "best.pt" file once the training is complete. This file incorporates both the configuration parameters as well as the adjusted weights based on your training data. Therefore, there's no necessity to add, create, or download any other cfg files or weight files for this stage.

Notably, if you're planning to process a large dataset (like the 15000 images you mentioned), you may want to consider switching to a variant like YOLOv8m, which is designed for larger datasets.

Hopefully this addresses your query, but don't hesitate to ask if you have any further questions!

**Comment by mgabell** [comment]
> Howdy!
> 
> Honestly you first part of the project looks fine and impressive! If you are doing 15000 images and detect, think of moving to yolov8m rather small.
> 
> The second part with AGN Onix, and I don't understand from that point.
> 
> > Comment: So yo my understanding. The code will load a pre-trained yolov8 model of type small. Then all needed is to train it and it will create a best.pt under results with the new updated model. There is no need at this stage to add. create or download any cfg files or weight files?
> 
> **Yeah, after training with your data using yolov8s.pt, you achieve a best.pt which carries cfg and weights. As mentioned, consider yolov8m**!
> 
> Hope this helps!

Thanks for answering. The Orin part I understand was confusing. It is to me too.
But in principle there are two ways of doing this. 
+"marcoslucianops" version of Deepstream-Yolo and his export. Following his instruction you are not supposed to export the weights and cfg when building the library. Just build it and then convert using export_yolov8.py
+ Seeeds version says that you need to export the cfg and weights from the pt and build with these files in the build directory. I interpret that they will use the files for the build somehow.

So. my original problem is that when running the nVidia test app deepstream-test1 with yolov8 I convert the model to onnx and according to marcoslucianops version. This has worked for me once or twice (building engine etc.). It always works with the coco yolov8.pt downloaded from ultralytics repo and convert to ONNX. But training my own and convert to ONNX and running I have no detections. It is blank. Even though it works fine using the ultralytics predict method. So somewhere in the conversion to ONNX it goes wrong. I do not get error messages anywhere which bugs me.

**Comment by mgabell** [comment]
> @mgabell hello!
> 
> You've grasped it correctly! When you train using the pretrained YOLOv8 model (like yolov8s.pt), it will generate a "best.pt" file once the training is complete. This file incorporates both the configuration parameters as well as the adjusted weights based on your training data. Therefore, there's no necessity to add, create, or download any other cfg files or weight files for this stage.
> 
> Notably, if you're planning to process a large dataset (like the 15000 images you mentioned), you may want to consider switching to a variant like YOLOv8m, which is designed for larger datasets.
> 
> Hopefully this addresses your query, but don't hesitate to ask if you have any further questions!

Have you benchmarked the accuracy in the models somehow? I try to run YoloV8 and compare to my original resnet50 model. I take the 40'000+ images and run inference on the resnet50 results with the YoloV8. In best of worlds they should come up with the same number, meaning that all images should have a hit and the number of objects be the same for each class. Of course the resnet50 can be wrong, and that I am trying to investigate now with the output from each to see how good they perform.

**Comment by mgabell** [comment]
For my application, starting with the yolov8m model you provide. Can I get better performance choosing another start model?
I was thinking that using a model trained on Coco may not be correct for my application. I am not looking for people and cars but more product defects.

**Comment by PiyushKamboj1503** [comment]
> For my application, starting with the yolov8m model you provide. Can I get better performance choosing another start model? I was thinking that using a model trained on Coco may not be correct for my application. I am not looking for people and cars but more product defects.

you can use the architecture by simply changing the model from 'yolov8m.pt' to 'yolov8m.yaml'.

**Comment by mgabell** [comment]
> > For my application, starting with the yolov8m model you provide. Can I get better performance choosing another start model? I was thinking that using a model trained on Coco may not be correct for my application. I am not looking for people and cars but more product defects.
> 
> you can use the architecture by simply changing the model from 'yolov8m.pt' to 'yolov8m.yaml'.

this would mean that I start from scratch. But I was more thinking that starting with different type of images in the pretrained model: https://captain-whu.github.io/AID/
They look more like spots and defects? If I start with a model trained on people I would assume transfer learning based on that for spots and scratches would be less accurate?

**Comment by zshariff6506** [comment]
> > For my application, starting with the yolov8m model you provide. Can I get better performance choosing another start model? I was thinking that using a model trained on Coco may not be correct for my application. I am not looking for people and cars but more product defects.
> 
> you can use the architecture by simply changing the model from 'yolov8m.pt' to 'yolov8m.yaml'.

Don't your .pt files actually do something, your .yaml models just have classes that the machine learning model detects.

**Comment by PiyushKamboj1503** [comment]
I hope the author finds these useful insights

**Comment by glenn-jocher** [comment]
@mgabell hello,

Thank you for taking the time to engage with the YOLOv8 repository. It's always valuable to hear user feedback and insights.

For your query about leveraging a different pre-trained model, it's certainly feasible. If the pre-trained model is trained on an image dataset that is more similar to your task (product defect detection), it could potentially yield better performance. This approach is based on the premise that the pre-trained model has learned more relevant features for your specific task during its initial training. 

However, keep in mind that the transfer learning method is multifaceted and involves more than just the utilisation of pre-trained models. Aspects such as fine-tuning on your specific data and appropriate selection of architecture (for example, choosing between 'yolov8m.pt' and 'yolov8m.yaml') are also crucially important to achieve optimal results.

The "pt" files refer to pre-trained PyTorch models which carry the weights learned from the pre-training, while "yaml" files denote the structure and hyperparameters of the model. When you're switching between these models, it's necessary to keep in mind what these individual components deliver and how they contribute to your use-case. 

We'd be happy to continue helping and support you further with any other queries you have as you advance your work!

Best of luck with your project!

**Comment by mgabell** [comment]
> @mgabell hello,
> 
> Thank you for taking the time to engage with the YOLOv8 repository. It's always valuable to hear user feedback and insights.
> 
> For your query about leveraging a different pre-trained model, it's certainly feasible. If the pre-trained model is trained on an image dataset that is more similar to your task (product defect detection), it could potentially yield better performance. This approach is based on the premise that the pre-trained model has learned more relevant features for your specific task during its initial training.
> 
> However, keep in mind that the transfer learning method is multifaceted and involves more than just the utilisation of pre-trained models. Aspects such as fine-tuning on your specific data and appropriate selection of architecture (for example, choosing between 'yolov8m.pt' and 'yolov8m.yaml') are also crucially important to achieve optimal results.
> 
> The "pt" files refer to pre-trained PyTorch models which carry the weights learned from the pre-training, while "yaml" files denote the structure and hyperparameters of the model. When you're switching between these models, it's necessary to keep in mind what these individual components deliver and how they contribute to your use-case.
> 
> We'd be happy to continue helping and support you further with any other queries you have as you advance your work!
> 
> Best of luck with your project!

Thanks! I am wiser 👍 
The challenge is to "easily" tune parameters so that we can get the best model for our application. No shortcuts here.

Thank you for your answers! Appriciated.
Only part left after getting the model to work with Orin deepstream.

I noticed in the past that you relied a lot on openCV. I have always had a hunch that a lot of the features in the nVidia hardware is bypassed if not using deepstream. Alot of CPU overhead and memory copying instead of using the GPU and its cache memories to its fullest. Is this something you have been able to overcome with the latest version of ultralytics? I have had a 
feeling that this is making ultralytics approach slower than deepstream. BUT ultralytics is easier to get started....

**Comment by glenn-jocher** [comment]
Hello @mgabell,

We're happy to hear that you found our previous explanations useful!

Indeed, fine-tuning parameters to get the most optimal model for a specific application can be an intricate process. There are no shortcuts, but the effort invested usually leads to a lot more robust and efficient models.

Coming to your query about OpenCV, you're right in noting that an over-reliance on CPU-based processing frameworks like OpenCV can lead to inefficiencies when more powerful GPU resources are available. One of our subsequent focuses in the newer iterations of the YOLOv8 is indeed to optimize the hardware-software interaction better. 

While Deepstream allows for a very GPU-centric workflow, it might not be as flexible or versatile for a broad range of tasks as some researchers might prefer. We do constantly strive to strike a balance by ensuring that while our platform remains beginner-friendly, it should not compromise on the efficiency and performance fronts. 

Our latest versions have seen improvements in both aspects, and we'll continue to learn from user feedback like yours and the broader computer vision community to iterate and improve. 

Thank you again for your feedback and don't hesitate to reach out if you have more questions or suggestions as you proceed with your project! We're here to support you.
